Make DataValueDeserializer accept callbacks

thiemowmde opened the pull request:
Not an alternative but an addition to https://github.com/wmde/WikibaseDataModel/pull/669.

[Bug: T132592](https://phabricator.wikimedia.org/T132592)

JeroenDeDauw commented:
I'd have added a second constructor parameter, though this also works. +1 to the approach

thiemowmde commented:
I finished this patch, rewrote the most critical bits of documentation and removed the WIP. Please review carefully and merge. This should be a simple addition, nothing should break.

brightbyte commented on src/Deserializers/DataValueDeserializer.php:
```diff
@@ -15,45 +16,56 @@
+	 *  DataValue objects, or full qualified DataValue class names.
 	 */
-	private $classes;
+	private $builders;
```
"builder" is not a good term for this. $callbacks? $handlers?

JeroenDeDauw replied:
I think $handlers is a lot worse than $builders, and find $callbacks less intent revealing

Really does not matter much IMO tho

brightbyte commented on src/Deserializers/DataValueDeserializer.php:
```diff
@@ -15,45 +16,56 @@
 	/**
-	 * @var string[] Associative array of data type id => DataValue class name.
+	 * @var array Associative array mapping data type IDs to either callables returning new
+	 *  DataValue objects, or full qualified DataValue class names.
```
What do the callbacks get as a parameter? If it's just a class name, what are the expectations? I assume such classes must have a static newFromArray method?
